### ISSUE TITLE
Zappa is ignoring some files and modules from installed packages

### ISSUE THREAD

**Issue by eldos-dl** [Miserlou/Zappa#1359]
## Context
I've noticed that Zappa is ignoring certain files during packaging. I believe this issue is caused as we are trying to exclude packages already builtin to the python lambda environment namely
```
 exclude = [
    "boto3",
    "dateutil",
    "botocore",
    "s3transfer",
    "six.py",
    "jmespath",
    "concurrent"
]
```
If any of the other python packages have any of the files or subfolders with the above names those are getting excluded as well. I noticed this as I was getting an `ImportError` while trying to do this `from django.utils import six`
 
## Expected Behavior
The above mentioned patterns should only be ignored if they are in the root level virtualenv's `site-packages` but not as sub-module of some other package.
## Actual Behavior
These patterns are being excluded at all levels. 

## Possible Fix
Copying everything and removing these files in a later step or updating the patterns.

## Steps to Reproduce
1. Install `django==1.11.6`, `zappa==0.45.1` in virtualenv
2. Create some zappa configuration without exclude argument
3. Ensure that there are no compiled files. (If there are compiled files `from django.utils import six` will not show an error as `six.pyc` would be available.)
3. Run `zappa package env`
4. Inspect the package for file `django/utils/six.py`

## Environment
* Zappa version used: 0.45.1
* Operating System and Python version: Ubuntu 16.04, Python 2.7.12

**Comment by suriya** [comment]
Zappa uses Python's [`copytree`](https://docs.python.org/2/library/shutil.html#shutil.copytree) with the `ignore` parameter constructed using [`shutil.ignore_patterns`](https://docs.python.org/2/library/shutil.html#shutil.ignore_patterns). This mechanism is not very user-friendly.

Ideally, if could support `gitignore` style patters or regular expressions that would be very convenient. @Miserlou Your thoughts?

**Comment by scoates** [comment]
See also #1343

**Comment by scoates** [comment]
Adding this for reference: https://pypi.python.org/pypi/pathspec

**Comment by scoates** [comment]
Related: #1083

**Comment by suriya** [comment]
I would like to submit a PR to address this issue. My commit is here: https://github.com/suriya/Zappa/commit/61d2972dd9a303440b1f155733bb4ee8ef220a59

The commit adds a configuration option called `exclude_pathspec` that specifies what files should be excluded from the final ZIP package. `exclude_pathspec` is a list of `.gitignore` style specifications. This is in addition to the existing `exclude` configuration. Having this as an additional option ensures backward compatibility. I have found that `exclude` has to be used with extreme care. At some point in the distant future, we could deprecate and remove `exclude`.

The commit adds a dependency, the pathspec library https://pypi.python.org/pypi/pathspec 

@Miserlou @scoates Please help me make this a pull request.

I don't like the configuration name `exclude_pathspec`. I prefer the name `gitignore`. If you have any suggestions, I could use that. I will update the documentation as well.

I could not find a test for `exclude`. Maybe it would be to good to have a test for `exclude` and `exclude_pathspec`.

**Comment by suriya** [comment]
@Miserlou @scoates Could you please take a look at https://github.com/suriya/Zappa/commit/61d2972dd9a303440b1f155733bb4ee8ef220a59 and help get this change into Zappa?

**Comment by TMiguelT** [comment]
@suriya this would be a very useful change. Perhaps you could make a formal PR so we could get it merged?

**Comment by suriya** [comment]
@TMiguelT I can make a PR. Before doing so, would like to get feedback on [my comment above](https://github.com/Miserlou/Zappa/issues/1359#issuecomment-370767007). Is the configuration name good, or do you have a better suggestion? How to add tests for this?

**Comment by TMiguelT** [comment]
> exclude_pathspec is a list of .gitignore style specifications.

Yes, `.gitignore` format is a good idea because people are familiar with it, and because it handles relevant edge cases quite well (like un-ignoring subfolders)

> This is in addition to the existing exclude configuration. Having this as an additional option ensures backward compatibility

I agree with this, I doubt we want to break existing configurations. Unless anyway (@Miserlou?) would rather a breaking change 

> I don't like the configuration name exclude_pathspec. I prefer the name gitignore. If you have any suggestions, I could use that. I will update the documentation as well.

I think this name is fine. It describes how it's different from `exclude`. `gitignore` doesn't really make sense because this isn't applying to a git repository

>  How to add tests for this?

I agree some tests would be helpful here. The way I would do this is by writing a test for the `zappa.utilities.copytree` function. Copy some files into a temporary directory (using `tempfile`), and verify that it has copied the right files, using a variety of different options (`ignore`, and `ignore_pathspec`)

I think your implementation looks fine as well, I doubt anyone would have a problem with this.

**Comment by felix-zg** [comment]
Any update on this?

**Comment by suriya** [comment]
Dear Maintainers: Could you please merge pull request https://github.com/Miserlou/Zappa/pull/2070 and close this issue?

**Comment by dfpoon** [comment]
I am not sure whether here is the right place to report this issue. But it took me a few hours to identify it. Hope this message will help others. 

Problem: When some unused .py files are listed in the "exclude" line zappa_settings.json, the AWS lambda will report weird problems. Here is the sample:

```
zappa_settings.json:
"exclude": [xxx.py, yyy.py],

```
When Zappa completed the 'update' rountine, the console reports the following error:

`Error: Warning! Status check on the deployed lambda failed. A GET request to '/' yielded a 502 response code.`

When I check the AWS CloudWatch Log, it shows: 
`Runtime.ImportModuleError: Unable to import module 'handler': No module named 'werkzeug.test'
`

This weird error is totally unrelated to the "xxx.py" and "yyy.py" at all. Besides, "xxx.py" and "yyy.py" are just test cases which are not involved in the HTTP server codes. The "import" package lines have no mentions of Werkzeug either.

My wild guess is that the "exclude" setting has removed some essential packages due to those .py files. Please investigate and fix the problem. Thanks.

**Comment by suriya** [comment]
Could someone help merge pull request #2070 and close this issue? I don't know how to make the perfect pull request. I feel this feature is useful to this project. If someone could polish and merge it, I would be much grateful. @TMiguelT

**Comment by TMiguelT** [comment]
I don't have any power to merge, unfortunately. I was just helping your PR with a first-pass review. I'm afraid we'll have to wait for the maintainers to get around to your PR. Perhaps consider asking on the Slack channel?

**Comment by stryker-sk** [comment]
This is an issue when trying to install Sentry. The following error exists: `ModuleNotFoundError: No module named 'sentry_sdk.integrations.boto3'`
